File: bulk-update-stacks/python/update_stacks.py
import os
import logging
import json
import urllib.request


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

CONFIG = {
    "spacelift_url": f"https://{os.environ.get("SPACELIFT_DOMAIN")}.app.spacelift.io/graphql" if os.environ.get("SPACELIFT_DOMAIN", False) else logger.error("SPACELIFT_DOMAIN environment variable is not set. Please set it to your Spacelift domain."),
    "spacelift_api_key_id": f"{os.environ.get("SPACELIFT_API_KEY_ID")}" if os.environ.get("SPACELIFT_API_KEY_ID", False) else logger.error("SPACELIFT_API_KEY_ID environment variable is not set. Please set it to your Spacelift API Key ID."),
    "spacelift_api_key_secret": f"{os.environ.get("SPACELIFT_API_KEY_SECRET")}" if os.environ.get("SPACELIFT_API_KEY_SECRET", False) else logger.error("SPACELIFT_API_KEY_SECRET environment variable is not set. Please set it to your Spacelift API Key Secret."),

    # update this object with whatever settings you'd like to update.

    # example attributes:

    # "default_branch": "main",
    # "namespace": "your-github/gitlab namespace",  # e.g. "spacelift"
    # if using public worker pool, you can ignore this field
    # "worker_pool": "your-worker-pool-id",
    # "provider": "GITHUB_ENTERPRISE",  # Or GITLAB
    # "terraform_version": "1.9.1"
}

QUERIES = {
    "get_stacks": """
            query {
                stacks{
                id
                }
            }
            """,
    "get_stack": """
            query($id:ID!) {
                stack(id: $id) {
                    repository
                    branch
                    projectRoot
                    description
                    space
                    administrative
                    namespace
                    provider
                    labels
                    terraformVersion
                    workerPool {
                        id
                        space	
                    }
                    vcsIntegration {
                        id
                    }
                }
            }
            """,
    "update_stack": """
            mutation UpdateStack($stackId: ID!, $input: StackInput!) {
                stackUpdate(id: $stackId, input: $input) {
                    id
                    __typename
                }
            }
            """,
    "get_jwt": """
        mutation GetSpaceliftToken($keyId: ID!, $keySecret: String!) {
            apiKeyUser(id: $keyId, secret: $keySecret) {
                id
                jwt
            }
        }
        """
}


class CustomError:
    def __init__(self, description, dump):
        self.description = description
        self.dump = dump

    def __call__(self, e=None):
        error_message = f"""
{self.description}


{json.dumps(self.dump, indent=2)}
        """
        print(error_message)
        if e is not None:
            print(e)
        raise Exception(f"{self.description}")


def query_api(query: str, check_data=None, variables: dict = None, err: CustomError = None) -> dict:
    try:
        if "GetSpaceliftToken" in query:
            headers = {
                "Content-Type": "application/json",
            }
        else:
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {get_spacelift_auth_token()}"
            }

        data = {
            "query": query,
        }

        if variables is not None:
            data["variables"] = variables

        if data is not None:
            req = urllib.request.Request(
                CONFIG["spacelift_url"], json.dumps(data).encode('utf-8'), headers)

        with urllib.request.urlopen(req) as response:
            resp = json.loads(response.read().decode('utf-8'))

        if "errors" in resp:
            err = CustomError(
                description="GraphQL request failed",
                dump=resp["errors"][0]["message"] if resp["errors"] else resp
            )

        if check_data is not None:
            def ensure_data(_resp, check):
                # Create an array of keys
                keys = [s for s in check.split(".?") if s]
                if len(keys) == 0:
                    return

                # Ensure the first key exists in the response
                this_key = keys[0]
                if this_key not in _resp:
                    err.code = f"{err.code}_{this_key.upper()}_MISSING"
                    err.dump = resp  # Keep this as the main resp for the error
                    err(query)
                # Delete the first key from the array and call ensure_data again
                del keys[0]
                ensure_data(_resp[this_key], '.?'.join(keys))
            ensure_data(resp, check_data)
        return resp
    except Exception as e:
        err(e)


def get_spacelift_auth_token():
    API_KEY_ID = CONFIG["spacelift_api_key_id"]
    API_KEY_SECRET = CONFIG["spacelift_api_key_secret"]

    variables = {
        "keyId": API_KEY_ID,
        "keySecret": API_KEY_SECRET
    }

    response = query_api(QUERIES["get_jwt"], variables=variables)

    if response['data']['apiKeyUser']['jwt'] is not None:
        API_TOKEN = response['data']['apiKeyUser']['jwt']
        return API_TOKEN


def update_stacks():
    try:
        response = query_api(QUERIES["get_stacks"], check_data="data.?stacks", err=CustomError(
            "Failed to fetch stacks", None))

        stacks_list = [item["id"] for item in response["data"]["stacks"]]
        logger.info(f"Found {len(stacks_list)} stacks to update")
    except KeyError as e:
        logger.error(f"Unexpected response structure: {e}")
        return

    # Update each stack
    successful_updates = 0
    failed_updates = 0

    for stack in stacks_list:
        try:
            # Get stack details
            stack_data = query_api(
                QUERIES["get_stack"], variables={"id": stack}, check_data="data.?stack", err=CustomError(
                    "Failed to fetch stack details", None))["data"]["stack"]
            # Build update variables
            variables = {
                "stackId": stack,
                "input": {
                    "administrative": stack_data["administrative"],
                    "branch": stack_data["branch"],
                    "name": stack,
                    "namespace": stack_data["namespace"],
                    "provider": stack_data["provider"],
                    "repository": stack_data["repository"],
                    "projectRoot": stack_data["projectRoot"],
                    "description": stack_data["description"],
                    "space": stack_data["space"],
                    "vcsIntegrationId": stack_data["vcsIntegration"]["id"] if stack_data["vcsIntegration"] else None,

                    # example of bulk updating all stacks to use a new worker pool:
                    "workerPool": CONFIG["worker_pool"],
                    "labels": stack_data["labels"],
                    "vendorConfig": {
                        "terraform": {
                            "version": stack_data["terraformVersion"],
                            "workflowTool": "OPEN_TOFU"
                        }
                    }
                }
            }

            # Update the stack
            logger.info(f"Updating stack: {stack}...")
            update_response = query_api(
                QUERIES["update_stack"], variables=variables, check_data="data.?stackUpdate", err=CustomError(
                    "Failed to update stack", None))
            # Check for errors in the update response
            if "errors" in update_response:
                logger.error(
                    f"Failed to update stack {stack}: {update_response['errors'][0]['message']}")
                failed_updates += 1
            else:
                logger.info(f"Successfully updated stack: {stack}")
                successful_updates += 1

        except Exception as e:
            logger.error(f"Failed to update stack {stack}: {e}")
            failed_updates += 1

    # Print summary
    logger.info(
        f"Update complete: {successful_updates} successful, {failed_updates} failed")

    if failed_updates > 0:
        logger.warning(
            f"{failed_updates} stacks failed to update. Check logs above for details.")


if __name__ == "__main__":
    update_stacks()

File: bulk-update-stacks/go/update_stacks.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"
)

// Config holds all configuration from environment variables
type Config struct {
	SpaceliftDomain       string
	SpaceliftAPIKeyID     string
	SpaceliftAPIKeySecret string
	SpaceliftURL          string
}

// NewConfig creates a new config from environment variables
func NewConfig() (*Config, error) {
	domain := os.Getenv("SPACELIFT_DOMAIN")
	keyID := os.Getenv("SPACELIFT_API_KEY_ID")
	keySecret := os.Getenv("SPACELIFT_API_KEY_SECRET")

	if domain == "" {
		return nil, fmt.Errorf("SPACELIFT_DOMAIN environment variable is required")
	}
	if keyID == "" {
		return nil, fmt.Errorf("SPACELIFT_API_KEY_ID environment variable is required")
	}
	if keySecret == "" {
		return nil, fmt.Errorf("SPACELIFT_API_KEY_SECRET environment variable is required")
	}

	config := &Config{
		SpaceliftDomain:       domain,
		SpaceliftAPIKeyID:     keyID,
		SpaceliftAPIKeySecret: keySecret,
		SpaceliftURL:          fmt.Sprintf("https://%s.app.spacelift.io/graphql", domain),
	}

	return config, nil
}

// GraphQL request structure
type GraphQLRequest struct {
	Query     string         `json:"query"`
	Variables map[string]any `json:"variables,omitempty"`
}

// GraphQL response structure
type GraphQLResponse struct {
	Data   any `json:"data"`
	Errors []struct {
		Message string `json:"message"`
	} `json:"errors"`
}

// Authentication response
type AuthResponse struct {
	Data struct {
		ApiKeyUser struct {
			ID  string `json:"id"`
			JWT string `json:"jwt"`
		} `json:"apiKeyUser"`
	} `json:"data"`
}

// Stack structures
type WorkerPool struct {
	ID    string `json:"id"`
	Space string `json:"space"`
}

type VCSIntegration struct {
	ID string `json:"id"`
}

type VendorConfig struct {
	Typename string `json:"__typename"`
}

type Stack struct {
	Administrative   bool            `json:"administrative"`
	ID               string          `json:"id"`
	Name             string          `json:"name"`
	Branch           string          `json:"branch"`
	Repository       string          `json:"repository"`
	RepositoryURL    string          `json:"repositoryURL"`
	ProjectRoot      string          `json:"projectRoot"`
	Space            string          `json:"space"`
	Namespace        string          `json:"namespace"`
	Provider         string          `json:"provider"`
	Labels           []string        `json:"labels"`
	Description      *string         `json:"description"`
	VCSIntegration   *VCSIntegration `json:"vcsIntegration"`
	TerraformVersion *string         `json:"terraformVersion"`
	WorkerPool       *WorkerPool     `json:"workerPool,omitempty"`
	VendorConfig     *VendorConfig   `json:"vendorConfig"`
}

type StacksResponse struct {
	Data struct {
		Stacks []Stack `json:"stacks"`
	} `json:"data"`
}

// Update result tracking
type UpdateResult struct {
	StackID string
	Success bool
	Error   error
}

type UpdateSummary struct {
	TotalStacks   int
	SuccessCount  int
	FailureCount  int
	SuccessStacks []string
	FailureStacks []UpdateResult
	ExecutionTime time.Duration
}

// SpaceliftClient handles GraphQL requests to Spacelift
type SpaceliftClient struct {
	token  string
	config *Config
	client *http.Client
}

func NewSpaceliftClient(config *Config) *SpaceliftClient {
	return &SpaceliftClient{
		config: config,
		client: &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *SpaceliftClient) authenticate() error {
	authQuery := `
		mutation GetSpaceliftToken($keyId: ID!, $keySecret: String!) {
			apiKeyUser(id: $keyId, secret: $keySecret) {
				id
				jwt
			}
		}
	`

	req := GraphQLRequest{
		Query: authQuery,
		Variables: map[string]interface{}{
			"keyId":     c.config.SpaceliftAPIKeyID,
			"keySecret": c.config.SpaceliftAPIKeySecret,
		},
	}

	var authResp AuthResponse
	if err := c.makeRequest(req, &authResp); err != nil {
		return fmt.Errorf("authentication failed: %w", err)
	}

	c.token = authResp.Data.ApiKeyUser.JWT
	return nil
}

func (c *SpaceliftClient) getAllStacks() ([]Stack, error) {
	query := `
		query {
			stacks {
				id
				name
				branch
				repository 
				repositoryURL
				labels
				projectRoot
				description
				space
				administrative
				namespace
				provider
				terraformVersion
				workerPool {
					id
					space
				}
				vcsIntegration {
					id
				}
				vendorConfig {
					__typename
				}
			}
		}
	`

	req := GraphQLRequest{Query: query}
	var resp StacksResponse
	if err := c.makeRequest(req, &resp); err != nil {
		return nil, fmt.Errorf("failed to get stacks: %w", err)
	}

	return resp.Data.Stacks, nil
}

func (c *SpaceliftClient) updateStack(stackID string, stack *Stack) error {
	updateQuery := `
		mutation UpdateStack($stackId: ID!, $input: StackInput!) {
			stackUpdate(id: $stackId, input: $input) {
				id
				__typename
			}
		}
	`

	// Preserve existing labels and add new one
	addNewLabels := make([]string, len(stack.Labels))
	copy(addNewLabels, stack.Labels)

	input := map[string]any{}

	// Administrative
	if val := os.Getenv("SPACELIFT_ADMINISTRATIVE"); val != "" {
		input["administrative"] = val
	} else {
		input["administrative"] = stack.Administrative
	}

	// Name
	if val := os.Getenv("SPACELIFT_NAME"); val != "" {
		input["name"] = val
	} else {
		input["name"] = stack.Name
	}

	// Branch
	if val := os.Getenv("SPACELIFT_BRANCH"); val != "" {
		input["branch"] = val
	} else {
		input["branch"] = stack.Branch
	}

	// Namespace
	if val := os.Getenv("SPACELIFT_NAMESPACE"); val != "" {
		input["namespace"] = val
	} else {
		input["namespace"] = stack.Namespace
	}

	// Provider
	if val := os.Getenv("SPACELIFT_PROVIDER"); val != "" {
		input["provider"] = val
	} else {
		input["provider"] = stack.Provider
	}

	// Repository
	if val := os.Getenv("SPACELIFT_REPOSITORY"); val != "" {
		input["repository"] = val
	} else {
		input["repository"] = stack.Repository
	}

	// RepositoryURL
	if val := os.Getenv("SPACELIFT_REPOSITORY_URL"); val != "" {
		input["repositoryURL"] = val
	} else {
		input["repositoryURL"] = stack.RepositoryURL
	}

	// ProjectRoot
	if val := os.Getenv("SPACELIFT_PROJECT_ROOT"); val != "" {
		input["projectRoot"] = val
	} else {
		input["projectRoot"] = stack.ProjectRoot
	}

	// Space
	if val := os.Getenv("SPACELIFT_SPACE"); val != "" {
		input["space"] = val
	} else {
		input["space"] = stack.Space
	}

	// Labels (comma-separated)
	if val := os.Getenv("SPACELIFT_LABELS"); val != "" {
		labels := strings.Split(val, ",")
		for i := range labels {
			labels[i] = strings.TrimSpace(labels[i])
		}
		input["labels"] = append(addNewLabels, labels...)
	} else {
		input["labels"] = stack.Labels
	}

	// WorkerPool
	if workerPool := os.Getenv("SPACELIFT_WORKER_POOL_ID"); workerPool != "" {
		input["workerPool"] = workerPool
	} else if stack.WorkerPool != nil {
		input["workerPool"] = stack.WorkerPool.ID
	}
	req := GraphQLRequest{
		Query: updateQuery,
		Variables: map[string]any{
			"stackId": stackID,
			"input":   input,
		},
	}

	var resp GraphQLResponse
	return c.makeRequest(req, &resp)
}

func (c *SpaceliftClient) makeRequest(req GraphQLRequest, result interface{}) error {
	jsonData, err := json.Marshal(req)
	if err != nil {
		return err
	}

	httpReq, err := http.NewRequest("POST", c.config.SpaceliftURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}

	httpReq.Header.Set("Content-Type", "application/json")
	if c.token != "" {
		httpReq.Header.Set("Authorization", "Bearer "+c.token)
	}

	resp, err := c.client.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP error: %d, body: %s", resp.StatusCode, string(body))
	}

	// Check for GraphQL errors
	var graphqlResp GraphQLResponse
	if err := json.Unmarshal(body, &graphqlResp); err != nil {
		return err
	}

	if len(graphqlResp.Errors) > 0 {
		return fmt.Errorf("GraphQL error: %s", graphqlResp.Errors[0].Message)
	}

	return json.Unmarshal(body, result)
}

func updateStack(client *SpaceliftClient, stack Stack, resultChan chan<- UpdateResult, wg *sync.WaitGroup) {
	defer wg.Done()

	fmt.Printf("Processing stack: %s\n", stack.ID)

	// Update the stack directly using the information from getAllStacks
	err := client.updateStack(stack.ID, &stack)
	result := UpdateResult{
		StackID: stack.ID,
		Success: err == nil,
		Error:   err,
	}

	if err == nil {
		fmt.Printf("✓ Successfully updated stack: %s\n", stack.ID)
	} else {
		fmt.Printf("✗ Failed to update stack %s: %v\n", stack.ID, err)
	}

	resultChan <- result
}

func main() {
	startTime := time.Now()

	// Load configuration from environment variables
	config, err := NewConfig()
	if err != nil {
		log.Fatalf("Configuration error: %v", err)
	}

	// Create client and authenticate
	client := NewSpaceliftClient(config)

	fmt.Println("Authenticating with Spacelift...")
	if err := client.authenticate(); err != nil {
		log.Fatalf("Authentication failed: %v", err)
	}
	fmt.Println("✓ Authentication successful")

	// Get all stacks
	fmt.Println("Retrieving all stacks...")
	stacks, err := client.getAllStacks()
	if err != nil {
		log.Fatalf("Failed to get stacks: %v", err)
	}
	fmt.Printf("✓ Retrieved %d stacks\n", len(stacks))

	// Start updating stacks concurrently
	fmt.Println("Starting bulk update...")
	var wg sync.WaitGroup
	resultChan := make(chan UpdateResult, len(stacks))

	// Launch goroutine for each stack
	terraformStacks := 0
	for _, stack := range stacks {
		if stack.VendorConfig.Typename != "StackConfigVendorTerraform" {
			continue
		}
		terraformStacks++
		wg.Add(1)
		go updateStack(client, stack, resultChan, &wg)
	}

	// Wait for all updates to complete
	wg.Wait()
	close(resultChan)

	// Collect results
	summary := UpdateSummary{
		TotalStacks:   terraformStacks,
		SuccessCount:  0,
		FailureCount:  0,
		FailureStacks: make([]UpdateResult, 0),
		ExecutionTime: time.Since(startTime),
	}

	for result := range resultChan {
		if result.Success {
			summary.SuccessCount++
		} else {
			summary.FailureCount++
			summary.FailureStacks = append(summary.FailureStacks, result)
		}
	}

	// Print summary
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("BULK UPDATE SUMMARY")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("Total stacks processed: %d\n", summary.TotalStacks)
	fmt.Printf("Successful updates: %d\n", summary.SuccessCount)
	fmt.Printf("Failed updates: %d\n", summary.FailureCount)
	fmt.Printf("Execution time: %v\n", summary.ExecutionTime)
	fmt.Printf("Success rate: %.2f%%\n", float64(summary.SuccessCount)/float64(summary.TotalStacks)*100)

	if summary.FailureCount > 0 {
		fmt.Println("\n✗ Failed to update stacks:")
		for _, failure := range summary.FailureStacks {
			fmt.Printf("  - %s: %v\n", failure.StackID, failure.Error)
		}
	}

	fmt.Println(strings.Repeat("=", 60))
}
